tests/switchless/host/host.c

```diff
@@ -0,0 +1,234 @@
+{
+    LARGE_INTEGER rval;
+    rval.QuadPart = stop.QuadPart - start.QuadPart;
+    return rval;
+} /* elapsed */
+
+static double timespan_to_sec(oe_timer_t ts)
+{
+    LARGE_INTEGER freq;
+    QueryPerformanceFrequency(&freq);
+    return (double)ts.QuadPart / (double)freq.QuadPart;
+} /* timespan_to_sec */
+
+#elif defined __GNUC__
+
+static long NS_PER_SEC = 1000000000;
+
+typedef struct timespec oe_timer_t;
+
+static void get_time(oe_timer_t* p_time)
+{
+    clock_gettime(CLOCK_REALTIME, p_time);
+} /* get_time */
+
+static oe_timer_t elapsed(oe_timer_t start, oe_timer_t stop)
+{
+    oe_timer_t rval = {0, 0};
+    if (start.tv_nsec > stop.tv_nsec)
+    {
+        rval.tv_sec = stop.tv_sec - (start.tv_sec + 1);
+        rval.tv_nsec = NS_PER_SEC + stop.tv_nsec - start.tv_nsec;
+    }
+    else
+    {
+        rval.tv_sec = stop.tv_sec - start.tv_sec;
+        rval.tv_nsec = stop.tv_nsec - start.tv_nsec;
+    }
+    return rval;
+} /* elapsed */
+
+static double timespan_to_sec(oe_timer_t ts)
+{
+    return (double)ts.tv_nsec / (double)NS_PER_SEC + (double)ts.tv_sec;
+} /* timespan_to_sec */
+
+#endif /*_MSC_VER or __GNUC__ */
+
+enum
+{
+    SAMPLE_SIZE = 2048
+};
+
+static int generate_random_number()
+{
+    return rand() % (INT_MAX / 2);
+} /* generate_random_number */
+
+static double test_standard_enc_sum(oe_enclave_t* enclave)
+{
+    oe_timer_t start, stop;
+    addition_args* args =
+        (addition_args*)malloc(sizeof(addition_args) * SAMPLE_SIZE);
+
+    /* generate sample data */
+    for (size_t i = 0; i < SAMPLE_SIZE; ++i)
+    {
+        args[i].arg1 = generate_random_number();
+        args[i].arg2 = generate_random_number();
+        args[i].sum = 0;
+    }
+
+    /* record start time */
+    get_time(&start);
+
+    for (size_t i = 0; i < SAMPLE_SIZE; ++i)
+    {
+        OE_TEST(
+            OE_OK == standard_enc_sum(
+                         enclave, &(args[i].sum), args[i].arg1, args[i].arg2));
+    }
+
+    /* record stop time */
+    get_time(&stop);
+
+    /* verify results */
+    for (size_t i = 0; i < SAMPLE_SIZE; ++i)
+    {
+        OE_TEST(args[i].arg1 + args[i].arg2 == args[i].sum);
+    }
+
+    free(args);
+
+    return timespan_to_sec(elapsed(start, stop));
+} /* test_standard_enc_sum */
+
+static double test_synchronous_switchless_enc_sum(oe_enclave_t* enclave)
+{
+    oe_timer_t start, stop;
+    addition_args* args =
+        (addition_args*)malloc(sizeof(addition_args) * SAMPLE_SIZE);
+
+    /* generate sample data */
+    for (size_t i = 0; i < SAMPLE_SIZE; ++i)
+    {
+        args[i].arg1 = generate_random_number();
+        args[i].arg2 = generate_random_number();
+        args[i].sum = 0;
+    }
+
+    /* explicitly start the enclave worker thread */
+    oe_switchless_manager_startup(enclave);
+
+    /* record start time */
+    get_time(&start);
+
+    for (size_t i = 0; i < SAMPLE_SIZE; ++i)
+    {
+        OE_TEST(
+            OE_OK == synchronous_switchless_enc_sum(
```

Comment:
The way we are testing now is even though this call is in a loop, we would enqueue the job and wait until the dequeued by the enclave worker thread. Essentially, this means the length of the queue is always <= 1. We should try to use more than one thread to call into this function to stress testing the queue implementation.

Reply:
There are a suite of tests explicitly for the queue which provide multi-producers testing.  I will also add a test to use multiple producers to the switchless test.

Reply:
Captured in issue #2037

Reply:
These tests have been added.